## 실습 1.

In [9]:
import pandas as pd
DATA_PATH = "book_bestseller_clean.csv"
df_books = pd.read_csv(DATA_PATH, encoding="utf-8-sig")
print("데이터 크기:", df_books.shape)
print("컬럼:", df_books.columns.tolist())
print("상품명 결측치:", df_books["상품명"].isna().sum())
df_books[["상품명"]].head(10)

데이터 크기: (199, 8)
컬럼: ['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야']
상품명 결측치: 0


,상품명
0,소년이 온다
1,모순
2,결국 국민이 합니다
3,혼모노
4,급류
5,초역 부처의 말
6,청춘의 독서(특별증보판)
7,어른의 행복은 조용하다
8,채식주의자
9,단 한 번의 삶(강물에디션 활판인쇄 한정판)


## 📊 `book_bestseller_clean.csv` 데이터 분석 결과 요약

### 1. 데이터 기본 구조
- **데이터 크기 (Shape):** 199행 × 8열
- **전체 컬럼 목록 (8개):** 
  - `순위`, `판매상품ID`, `상품명`, `판매가`, `저자`, `출판사`, `발행일`, `분야`

---

### 2. 주요 컬럼 검증 (상품명)
- **상품명 결측치 (Missing Values):** 0개 (결측치 없음)

---

### 3. 상품명 상위 10개 목록 (Sample Data)

| 번호 | 상품명 |
| :---: | :--- |
| 1 | 소년이 온다 |
| 2 | 모순 |
| 3 | 결국 국민이 합니다 |
| 4 | 혼모노 |
| 5 | 급류 |
| 6 | 초역 부처의 말 |
| 7 | 청춘의 독서(특별증보판) |
| 8 | 어른의 행복은 조용하다 |
| 9 | 채식주의자 |
| 10 | 단 한 번의 삶(강물에디션 활판인쇄 한정판) |

## 실습 2. 분석할 제목 문자열 준비하기

In [10]:
titles = (
    df_books["상품명"]
    .fillna("")
    .astype(str)
    .str.strip()
)

titles = titles[titles != ""].reset_index(drop=True)
print("사용할 제목 수:", len(titles))
titles.head(10)

사용할 제목 수: 199


0                      소년이 온다
1                          모순
2                  결국 국민이 합니다
3                         혼모노
4                          급류
5                    초역 부처의 말
6               청춘의 독서(특별증보판)
7                어른의 행복은 조용하다
8                       채식주의자
9    단 한 번의 삶(강물에디션 활판인쇄 한정판)
Name: 상품명, dtype: str

## 📌 `상품명` 컬럼 상위 10개 조회 결과 요약

- **자료형 (dtype):** `str` (문자열)
- **조회 범위:** Index `0` ~ `9` (상위 10개)

### 📋 도서 목록

| Index | 상품명 |
| :---: | :--- |
| 0 | 소년이 온다 |
| 1 | 모순 |
| 2 | 결국 국민이 합니다 |
| 3 | 혼모노 |
| 4 | 급류 |
| 5 | 초역 부처의 말 |
| 6 | 청춘의 독서(특별증보판) |
| 7 | 어른의 행복은 조용하다 |
| 8 | 채식주의자 |
| 9 | 단 한 번의 삶(강물에디션 활판인쇄 한정판) |

## 실습 3. 머신러닝은 왜 텍스트를 숫자로 바꿀까?

머신러닝 알고리즘은 일반적으로 숫자를 입력받아 계산합니다.

예를 들어 다음과 같은 숫자는 계산할 수 있습니다.


도서 제목

        ↓

사용된 단어 확인

        ↓

단어마다 열(column) 생성

        ↓

등장 여부 또는 등장 횟수를 숫자로 기록

        ↓

숫자 벡터

이 과정을 텍스트 벡터화(Vectorization)라고 합니다.

## 실습 4. 아주 작은 예제로 먼저 이해하기

실제 1,000권 정도의 도서 제목을 바로 사용하면 행과 열이 많아져 원리를 보기 어렵습니다.

In [11]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

# 1. 예제 텍스트 데이터 정의
sample_docs = [
    "파이썬 데이터 분석",
    "파이썬 머신러닝",
    "데이터 분석 입문",
]

# --------------------------------------------------
# [방법 1] CountVectorizer: 단어의 등장 빈도수(Count) 기준
# --------------------------------------------------
cv = CountVectorizer()
cv_matrix = cv.fit_transform(sample_docs)

# 결과를 보기 쉽게 DataFrame으로 변환
df_cv = pd.DataFrame(
    cv_matrix.toarray(), 
    columns=cv.get_feature_names_out(),
    index=[f"문장 {i+1}" for i in range(len(sample_docs))]
)

print("=== 1. CountVectorizer 결과 (단어 빈도수) ===")
print(df_cv)
print("\n" + "="*50 + "\n")

# --------------------------------------------------
# [방법 2] TfidfVectorizer: 단어의 중요도(TF-IDF) 기준
# --------------------------------------------------
tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(sample_docs)

# 결과를 보기 쉽게 DataFrame으로 변환
df_tfidf = pd.DataFrame(
    tfidf_matrix.toarray().round(3),  # 소수점 3자리까지 반올림
    columns=tfidf.get_feature_names_out(),
    index=[f"문장 {i+1}" for i in range(len(sample_docs))]
)

print("=== 2. TfidfVectorizer 결과 (단어 중요도) ===")
print(df_tfidf)

=== 1. CountVectorizer 결과 (단어 빈도수) ===
      데이터  머신러닝  분석  입문  파이썬
문장 1    1     0   1   0    1
문장 2    0     1   0   0    1
문장 3    1     0   1   1    0


=== 2. TfidfVectorizer 결과 (단어 중요도) ===
        데이터   머신러닝     분석     입문    파이썬
문장 1  0.577  0.000  0.577  0.000  0.577
문장 2  0.000  0.796  0.000  0.000  0.605
문장 3  0.518  0.000  0.518  0.681  0.000


## 실습 5. Bag of Words 이해하기

앞의 방식은 단어를 하나의 주머니에 넣고 몇 번 등장했는지 세는 것과 비슷합니다.

CountVectorizer (단어 빈도수)

원리: 단어가 몇 번 나왔는지만 카운트합니다.

한계: '은/는/이/가', '데이터', '분석'처럼 모든 문장에 자주 등장하는 일반적인 단어에도 높은 점수가 부여되는 단점이 있습니다.

TfidfVectorizer (단어 중요도 가중치)

원리: 단어 빈도(TF)에 역문서 빈도(IDF)를 곱해 "특정 문장에서만 유독 자주 쓰인 핵심 키워드"에 높은 점수를 줍니다.

장점: 문서의 주제나 특징을 대표하는 단어를 훨씬 잘 찾아냅니다.

## 실습 6. CountVectorizer 설치 확인하기

In [12]:
from sklearn.feature_extraction.text import CountVectorizer

## 실습 7. 작은 예제에 CountVectorizer 적용하기

In [17]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer

# 1. 예제 데이터 정의
sample_docs = [
    "파이썬 데이터 분석",
    "파이썬 머신러닝",
    "데이터 분석 입문",
]

# 2. CountVectorizer 객체 생성 및 학습/변환 (실습 7번)
count_vectorizer = CountVectorizer()
X_count_sample = count_vectorizer.fit_transform(sample_docs)

# 3. 결과 확인 (DataFrame으로 변환하여 보기 쉽게 출력)
df_count_sample = pd.DataFrame(
    X_count_sample.toarray(),
    columns=count_vectorizer.get_feature_names_out(),
    index=[f"문장 {i+1}" for i in range(len(sample_docs))]
)

print("=== CountVectorizer 변환 결과 ===")
print(df_count_sample)

=== CountVectorizer 변환 결과 ===
      데이터  머신러닝  분석  입문  파이썬
문장 1    1     0   1   0    1
문장 2    0     1   0   0    1
문장 3    1     0   1   1    0


## 실습 8. 생성된 단어 사전 확인하기

In [18]:
# 실습 8. 생성된 단어 사전 확인하기
feature_names = count_vectorizer.get_feature_names_out()

print("=== 생성된 단어 사전 (Feature Names) ===")
print(feature_names)

=== 생성된 단어 사전 (Feature Names) ===
['데이터' '머신러닝' '분석' '입문' '파이썬']


Plaintext
['데이터' '머신러닝' '분석' '입문' '파이썬']
get_feature_names_out(): CountVectorizer가 3개의 문장에서 추출한 중복 없는 단어들의 목록(배열)을 반환합니다.

단어들은 기본적으로 가나다순(알파벳순)으로 정렬되어 컬럼의 순서를 결정합니다.

## 실습 9. 단어-문서 행렬 확인하기

In [20]:
sample_count_df = pd.DataFrame(
    X_count_sample.toarray(),
    columns=feature_names,
    index=sample_docs,
)

sample_count_df

,데이터,머신러닝,분석,입문,파이썬
파이썬 데이터 분석,1,0,1,0,1
파이썬 머신러닝,0,1,0,0,1
데이터 분석 입문,1,0,1,1,0


{'파이썬': 2, '데이터': 1, '분석': 2, '머신러닝': 1, '입문': 1}

## 실습 10. 같은 단어가 여러 번 나오면 어떻게 될까?

다음 문장을 추가해 봅니다.

In [21]:
repeat_docs = [
    "파이썬 파이썬 데이터",
    "데이터 분석",
]

repeat_vectorizer = CountVectorizer()

X_repeat = repeat_vectorizer.fit_transform(repeat_docs)

repeat_df = pd.DataFrame(
    X_repeat.toarray(),
    columns=repeat_vectorizer.get_feature_names_out(),
    index=repeat_docs,
)
repeat_df

,데이터,분석,파이썬
파이썬 파이썬 데이터,1,0,2
데이터 분석,1,1,0


## 실습 11. CountVectorizer의 기본 토큰 기준 확인하기
CountVectorizer()는 기본 설정에서 문자열을 내부적으로 나누어 토큰을 만듭니다.

기본 토큰 패턴은 일반적으로 두 글자 이상의 단어 문자를 대상으로 합니다.

따라서 한 글자 토큰은 기본 설정에서 제외될 수 있습니다.

In [23]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer

# 1. 예제 데이터 정의 (한 글자 단어인 'R'과 'AI' 포함)
test_docs = [
    "AI 데이터 분석 R 파이썬",
]

# 2. 기본 설정의 CountVectorizer 적용
test_vectorizer = CountVectorizer()
X_test = test_vectorizer.fit_transform(test_docs)

# 3. 추출된 단어 목록 확인
print("=== 추출된 단어 목록 ===")
print(test_vectorizer.get_feature_names_out())

=== 추출된 단어 목록 ===
['ai' '데이터' '분석' '파이썬']


- **테스트 문장:** `"AI 데이터 분석 R 파이썬"`
- **추출 결과:** `['ai', '데이터', '분석', '파이썬']`
- **핵심 발견:**
  - 한 글자 토큰(`R`)은 기본 토큰 패턴(2글자 이상)에서 **자동 제외**됨
  - 대문자(`AI`)는 **소문자(`ai`)로 자동 변환**됨

## 실습 12. 실제 도서 제목에 CountVectorizer 적용하기

In [24]:
count_vectorizer = CountVectorizer()

X_count = count_vectorizer.fit_transform(titles)

count_terms = count_vectorizer.get_feature_names_out()

In [ ]:
print("문서 수:", X_count.shape[0])

print("단어 수:", X_count.shape[1])

print("행렬 크기:", X_count.shape)

print()

문서 수: 199
단어 수: 536
행렬 크기: (199, 536)


- **문서 수:** 199개
- **추출된 단어 수:** 536개
- **행렬 크기:** `(199, 536)` — 도서 제목마다 536차원 벡터로 표현

## 실습 13. vocabulary_ 확인하기 

In [26]:
# 실습 13. vocabulary_ 확인하기

# 단어 사전과 열 번호 매핑 정보 확인 (상위 20개 항목)
vocab_items = list(count_vectorizer.vocabulary_.items())[:20]

print("=== vocabulary_ 매핑 정보 (단어: 열 번호) ===")
print(vocab_items)

=== vocabulary_ 매핑 정보 (단어: 열 번호) ===
[('소년이', 278), ('온다', 355), ('모순', 194), ('결국', 67), ('국민이', 84), ('합니다', 511), ('혼모노', 524), ('급류', 92), ('초역', 460), ('부처의', 236), ('청춘의', 458), ('독서', 150), ('특별증보판', 485), ('어른의', 332), ('행복은', 517), ('조용하다', 432), ('채식주의자', 454), ('번의', 220), ('강물에디션', 57), ('활판인쇄', 528)]


- 단어와 열 번호를 매핑한 딕셔너리 (`{단어: 열 인덱스}`)
- 예시: `'소년이': 278`, `'온다': 355`, `'모순': 194`

## 실습 14. 희소 행렬(Sparse Matrix) 이해하기

In [27]:
# 실습 14. 희소 행렬(Sparse Matrix) 데이터 타입 확인 및 메모리 관리

# 1. Vectorizer 반환 데이터 타입 확인 (Scipy의 Sparse Matrix 타입 출력)
print("=== Vectorizer 반환 결과의 데이터 타입 ===")
print(type(X_count_sample))

# 2. 희소 행렬 내부의 0과 0이 아닌 값 비율 확인 (shape 및 nnz 활용)
print("\n=== 희소 행렬 정보 ===")
print("전체 행렬 크기:", X_count_sample.shape)
print("0이 아닌 실체 데이터 개수(nnz):", X_count_sample.nnz)

# 3. [주의] 전체 데이터(.toarray()) 대신 일부(상위 2개)만 Dense 배열로 확인하는 습관
print("\n=== 상위 2개 행만 추출해서 확인 ===")
print(X_count_sample[:2].toarray())

=== Vectorizer 반환 결과의 데이터 타입 ===
<class 'scipy.sparse._csr.csr_matrix'>

=== 희소 행렬 정보 ===
전체 행렬 크기: (3, 5)
0이 아닌 실체 데이터 개수(nnz): 8

=== 상위 2개 행만 추출해서 확인 ===
[[1 0 1 0 1]
 [0 1 0 0 1]]


- **반환 타입:** `scipy.sparse._csr.csr_matrix`
- **예제 행렬(3×5) 기준 0이 아닌 값 개수(nnz):** 8 / 15
- **실무 팁:** 전체를 `.toarray()`로 변환하면 메모리 낭비가 크므로, `X[:2].toarray()`처럼 **일부 행만 dense로 변환**해서 확인

## 실습 15. 실제 데이터의 첫 번째 제목 벡터 확인하기

In [28]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer

# 1. 실제 데이터에서 상품명(도서 제목) 추출 및 Vectorizer 학습
titles = df_books["상품명"]

count_vectorizer = CountVectorizer()
X_count = count_vectorizer.fit_transform(titles)
count_terms = count_vectorizer.get_feature_names_out()

# --------------------------------------------------
# [실습 15] 첫 번째 제목 벡터 및 추출된 단어 확인
# --------------------------------------------------

# 1) 원본 첫 번째 도서 제목 확인
print("=== 1. 원본 첫 번째 도서 제목 ===")
print(titles.iloc[0])
print("\n" + "="*50 + "\n")

# 2) 희소 행렬의 첫 번째 행(0번)에서 0보다 큰(실제 등장한) 단어만 추출
first_row = X_count.getrow(0)
indices = first_row.indices  # 단어가 위치한 컬럼 인덱스 번호들
values = first_row.data      # 해당 단어의 등장 횟수(값)들

first_title_terms = [
    (count_terms[index], value)
    for index, value in zip(indices, values)
]

print("=== 2. Vectorizer가 추출한 단어와 빈도수 ===")
print(first_title_terms)

=== 1. 원본 첫 번째 도서 제목 ===
소년이 온다


=== 2. Vectorizer가 추출한 단어와 빈도수 ===
[('소년이', np.int64(1)), ('온다', np.int64(1))]


- **원본 제목:** `"소년이 온다"`
- **`getrow(0).indices` / `.data`로 추출한 단어·빈도:** `[('소년이', 1), ('온다', 1)]`

## 실습 16. 전체 데이터에서 많이 등장한 단어 확인하기

In [29]:
import numpy as np
import pandas as pd

# 실습 16. 전체 데이터에서 많이 등장한 단어 확인하기

# 1. Count 행렬의 각 열(단어)별 합계 계산 (axis=0)
count_sums = np.asarray(X_count.sum(axis=0)).ravel()

# 2. 단어와 전체 등장 횟수를 데이터프레임으로 제작
count_summary = pd.DataFrame({
    "단어": count_terms,
    "전체등장횟수": count_sums,
})

# 3. 등장 횟수 기준 내림차순 정렬 후 인덱스 재설정
count_summary = count_summary.sort_values(
    "전체등장횟수",
    ascending=False,
).reset_index(drop=True)

# 4. 상위 30개 단어 출력
count_summary.head(30)

,단어,전체등장횟수
0,에디션,12
1,리커버,8
2,기념,8
3,위한,7
4,해커스,7
5,토익,6
6,나는,5
7,내가,5
8,스페셜,5
9,싶은,4


## 📊 `CountVectorizer` 전체 단어 등장 횟수 분석 (상위 30개)

### 1. 결과 요약
* **가장 높은 빈도의 단어:** `에디션` (12회)
* **주요 마케팅 키워드:** `에디션` (12회), `리커버` (8회), `기념` (8회), `스페셜` (5회), `한정판` (3회)
* **수험/수험서 키워드:** `해커스` (7회), `토익` (6회), `rc` (4회), `lc` (4회), `voca` (3회)
---

### 💡 주요 시사점 및 전처리 포인트
1. **불용어(Stopwords) 포함 확인:**
   * `위한`, `나는`, `내가`, `싶은`, `대하여`, `않는다`와 같은 조사/어미 성격의 단어들이 상위권에 포진해 있습니다.
   * 의미 있는 단어 위주로 분석하려면 **형태소 분석기(Kiwi 등)**를 통해 명사만 추출하거나 **불용어(Stopwords) 제거** 단계가 필요함을 알 수 있습니다.
2. **리패키지/마케팅 도서의 비중:**
   * `에디션`, `리커버`, `기념`, `한정판` 등 베스트셀러 도서 중 리패키지판이 상당수 포함되어 있음을 확인할 수 있습니다.

## 실습 17. Count 상위 단어 저장하기
상위 30개 단어를 저장합니다.

In [31]:
import pandas as pd

# 1. 상위 30개 단어 추출
count_top30 = count_summary.head(30)

# 2. CSV 파일로 저장 (한글 깨짐 방지를 위해 utf-8-sig 사용)
count_top30.to_csv(
    "chapter03_count_top_terms.csv",
    index=False,
    encoding="utf-8-sig",
)

# 3. 저장한 CSV 파일을 다시 불러와서 상위 5개 확인
pd.read_csv(
    "chapter03_count_top_terms.csv",
    encoding="utf-8-sig",
).head()

,단어,전체등장횟수
0,에디션,12
1,리커버,8
2,기념,8
3,위한,7
4,해커스,7


## 📌 `chapter03_count_top_terms.csv` 저장 및 불러오기 결과 (상위 5개)

`pd.read_csv()`를 통해 다시 읽어온 상위 5개 단어 확인 결과입니다.

### 📋 데이터 목록

| Index | 단어 | 전체등장횟수 |
| :---: | :--- | :---: |
| **0** | 에디션 | 12 |
| **1** | 리커버 | 8 |
| **2** | 기념 | 8 |
| **3** | 위한 | 7 |
| **4** | 해커스 | 7 |

## 실습 18. Count 방식의 한계 생각해 보기

CountVectorizer는 이해하기 쉽지만 한 가지 중요한 문제가 있습니다.

실습 18 핵심 개념 요약
CountVectorizer의 한계

모든 문서에 흔하게 등장하는 단어(예: 책, 도서, 이야기, 위한 등)도 등장 횟수가 많다는 이유만으로 높은 가중치(숫자)를 받게 됩니다.

이러한 보편적인 단어들은 정작 "이 문서가 어떤 특이한 내용을 다루고 있는가?"를 구별하는 데 아무런 도움을 주지 못합니다.

TF-IDF의 출발점

TF (Term Frequency, 단어 빈도): 특정 문서 안에서 해당 단어가 얼마나 많이 등장했는가?

IDF (Inverse Document Frequency, 역문서 빈도): 모든 문서 전체를 통틀어 해당 단어가 얼마나 희귀하게 등장했는가?

결론: 단순히 빈도수만 세는 것이 아니라, "흔한 단어의 점수는 깎고, 특정 문서에만 등장하는 유니크한 단어의 점수는 높이자"는 목적에서 TF-IDF가 탄생했습니다.

## 실습 19. TF 이해하기

습 19 핵심 개념 요약
TF (Term Frequency, 단어 빈도)

개념: 하나의 특정 문서(문장) 안에서 특정 단어가 얼마나 자주 등장하는지를 나타내는 값입니다.

예시: "파이썬 파이썬 데이터 분석" 문장에서 '파이썬'의 TF 값은 '데이터'나 '분석'보다 높습니다.

⚠️ scikit-learn TF-IDF 사용 시 필수 주의사항

scikit-learn의 TfidfVectorizer는 단순 빈도 계산에 더해 L2 정규화(Normalization) 및 IDF(역문서 빈도)를 곱해서 최종 수치를 만듭니다.

따라서 최종 출력되는 TF-IDF 수치(예: 0.587, 0.321)를 단순한 단어 등장 횟수로 직접 해석해서는 안 됩니다.

## 실습 20. DF 이해하기

 실습 20 핵심 개념 요약DF (Document Frequency, 문서 빈도)개념: 전체 문서 모음 중 특정 단어가 들어있는 문서의 개수를 의미합니다.예시 (전체 문서 100개 기준):'데이터': 80개 문서에 등장 $\rightarrow$ DF가 높음 (널리 쓰이는 흔한 단어)'약동학': 2개 문서에만 등장 $\rightarrow$ DF가 낮음 (특정 분야/주제에서만 쓰이는 희귀한 단어)DF의 역할과 의미DF는 특정 단어가 전체 문서 집합에 얼마나 널리 퍼져 있는지 알려주는 지표입니다.DF가 너무 높은 단어는 일반적인 단어일 확률이 높아 문서 분류나 키워드 추출 시 중요도가 낮게 평가되는 기준이 됩니다.

## 실습 21. IDF 이해하기

# 📚 Chapter 03. 텍스트를 숫자로 바꾸는 CountVectorizer와 TF-IDF

---

## 💡 TF-IDF 핵심 개념 요약

### 1. TF (Term Frequency, 단어 빈도)
* **개념:** 특정 문서 안에서 해당 단어가 얼마나 자주 등장하는지를 나타냅니다.
* **특징:** 하나의 문서 내부에서 단어의 중요도를 평가합니다.

### 2. DF (Document Frequency, 문서 빈도)
* **개념:** 전체 문서 모음 중 특정 단어가 포함된 문서의 개수를 의미합니다.
* **특징:** 단어가 전체 문서 집합에 얼마나 널리 퍼져 있는지를 보여줍니다.

### 3. IDF (Inverse Document Frequency, 역문서 빈도)
* **개념:** DF의 역수 개념으로, 단어의 희소성을 평가하는 가중치입니다.
* **동작 원리:**
  * **흔한 단어 (DF ↑):** 문서를 구분하는 능력이 떨어지므로 **IDF 점수 감점 (↓)**
  * **희귀한 단어 (DF ↓):** 특정 문서를 구분하는 핵심 키워드이므로 **IDF 점수 가산 (↑)**

---

## 📌 TF-IDF 계산식 및 역할

$$\text{TF-IDF} = \text{TF (문서 내 단어 빈도)} \times \text{IDF (전체 희소성 가중치)}$$

* **TF-IDF의 목적:** 단순히 빈도수만 세는 `CountVectorizer`의 한계를 극복하여, 흔한 단어의 영향력은 줄이고 **문서의 특징을 잘 나타내는 유니크한 키워드**에 더 높은 점수를 부여합니다.
* **주의사항:** `scikit-learn`은 L2 정규화 및 스무딩(Smoothing)이 적용되므로, 최종 출력값을 단순 등장 횟수로 직접 해석하지 않습니다.

## 실습 22. TF-IDF를 한 문장으로 정리하기

# 💡 실습 22. TF-IDF 핵심 요약 정리

---

### 📌 TF-IDF 한 문장 정의
> **"이 문서 내에서는 자주 등장하지만, 전체 문서 모음에서는 흔하지 않은 단어에 높은 가중치를 부여하는 지표"**

---

### 🔍 두 가지 핵심 관점의 결합

1. **TF (Term Frequency)**  
   * **질문:** *이 문서 안에서 얼마나 자주 나타나는가?*
   * **역할:** 해당 문서 내부에서의 단어 비중을 반영합니다.

2. **IDF (Inverse Document Frequency)**  
   * **질문:** *전체 문서 집합에서는 얼마나 흔하거나 드문가?*
   * **역할:** 모든 문서에 다 나오는 흔한 단어의 점수를 감점(가중치 축소)시킵니다.

---

### ⚠️ 해석 시 필수 주의사항

* **상대적 중요도:** TF-IDF 점수가 높다고 해서 현실 세계에서 무조건 절대적으로 중요한 단어라는 의미는 아닙니다.
* **조건부 특징:** **"현재 수집된 문서 집합(Corpus)"**과 **"설정한 Vectorizer 조건"** 내에서 상대적으로 차별성을 가지는 단어라는 뜻입니다.

## 실습 23. 작은 예제로 TfidfVectorizer 적용하기

In [32]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

# 1. 예제 데이터 및 기존 CountVectorizer 결과
sample_docs = [
    "파이썬 데이터 분석",
    "파이썬 머신러닝",
    "데이터 분석 입문",
]

# (비교용) CountVectorizer 적용
count_vectorizer = CountVectorizer()
X_count_sample = count_vectorizer.fit_transform(sample_docs)
sample_count_df = pd.DataFrame(
    X_count_sample.toarray(),
    columns=count_vectorizer.get_feature_names_out(),
    index=sample_docs,
)

# 2. [실습 23] TfidfVectorizer 적용
tfidf_sample_vectorizer = TfidfVectorizer()
X_tfidf_sample = tfidf_sample_vectorizer.fit_transform(sample_docs)
tfidf_sample_terms = tfidf_sample_vectorizer.get_feature_names_out()

tfidf_sample_df = pd.DataFrame(
    X_tfidf_sample.toarray(),
    columns=tfidf_sample_terms,
    index=sample_docs,
)

# 3. 결과 출력 및 비교
print("=== 1. CountVectorizer 결과 (정수형 빈도수) ===")
print(sample_count_df)

print("\n=== 2. TfidfVectorizer 결과 (소수점 3자리 반올림) ===")
print(tfidf_sample_df.round(3))

=== 1. CountVectorizer 결과 (정수형 빈도수) ===
            데이터  머신러닝  분석  입문  파이썬
파이썬 데이터 분석    1     0   1   0    1
파이썬 머신러닝      0     1   0   0    1
데이터 분석 입문     1     0   1   1    0

=== 2. TfidfVectorizer 결과 (소수점 3자리 반올림) ===
              데이터   머신러닝     분석     입문    파이썬
파이썬 데이터 분석  0.577  0.000  0.577  0.000  0.577
파이썬 머신러닝    0.000  0.796  0.000  0.000  0.605
데이터 분석 입문   0.518  0.000  0.518  0.681  0.000


## 📊 CountVectorizer vs TfidfVectorizer 결과 비교 분석

---

### 1. 주요 차이점 핵심 요약

* **값의 형태:** `Count`는 정수(등장 횟수)로 표현되는 반면, `TF-IDF`는 가중치 연산 및 L2 정규화가 적용되어 **소수**로 표현됩니다.
* **가중치 차등 부여:** `Count`에서는 모든 단어가 똑같이 `1` 점수를 받지만, `TF-IDF`에서는 **문서를 구분하는 독특한 단어일수록 더 높은 점수**를 받습니다.

---

### 2. 세부 분석: 왜 숫자가 다를까?

### ① "파이썬 머신러닝" 문장 분석
* **Count:** `'머신러닝'`과 `'파이썬'` 모두 동일하게 `1`로 평가됩니다.
* **TF-IDF:**
  * **`머신러닝` (0.796):** 전체 3개 문서 중 단 1곳에만 나오는 **희귀한 단어**이므로 가중치가 매우 높아집니다.
  * **`파이썬` (0.605):** 다른 문서에도 등장하는 흔한 단어이므로 희소성이 떨어져 상대적으로 낮게 평가됩니다.

### ② "데이터 분석 입문" 문장 분석
* **Count:** `'데이터'`, `'분석'`, `'입문'` 세 단어 모두 `1`로 동일합니다.
* **TF-IDF:**
  * **`입문` (0.681):** 이 문서에만 독점적으로 등장하므로 점수가 가장 높습니다.
  * **`데이터` (0.518), `분석` (0.518):** 첫 번째 문장에서도 함께 쓰인 흔한 단어들이라 점수가 깎입니다.

---

### 💡 결론
`TfidfVectorizer`를 사용하면 단순 횟수를 넘어 **"이 문장의 정체성을 잘 나타내는 핵심 단어가 무엇인가?"**를 수치화할 수 있습니다.

## 실습 24. TfidfVectorizer가 만든 단어 확인하기

In [33]:
# 실습 24. TfidfVectorizer가 만든 단어(컬럼 목록) 확인하기

# 1. Vectorizer가 학습하여 추출한 단어 목록 출력
feature_names = tfidf_sample_vectorizer.get_feature_names_out()

print("=== TfidfVectorizer가 추출한 단어 목록 ===")
print(feature_names)

=== TfidfVectorizer가 추출한 단어 목록 ===
['데이터' '머신러닝' '분석' '입문' '파이썬']


### 💻 실행 결과

```text
=== TfidfVectorizer가 추출한 단어 목록 ===
['데이터' '머신러닝' '분석' '입문' '파이썬']

## 실습 25. 단어별 IDF 값 확인하기

In [34]:
import pandas as pd

# 실습 25. 단어별 IDF 값 확인하기

# 1. 학습된 단어와 해당하는 IDF 값을 데이터프레임으로 생성
idf_df = pd.DataFrame({
    "단어": tfidf_sample_vectorizer.get_feature_names_out(),
    "IDF": tfidf_sample_vectorizer.idf_,
})

# 2. IDF 값 기준 내림차순(높은 순) 정렬
idf_df_sorted = idf_df.sort_values("IDF", ascending=False).reset_index(drop=True)

# 3. 결과 출력
print("=== 단어별 IDF 값 (내림차순 정렬) ===")
print(idf_df_sorted)

=== 단어별 IDF 값 (내림차순 정렬) ===
     단어       IDF
0  머신러닝  1.693147
1    입문  1.693147
2   데이터  1.287682
3    분석  1.287682
4   파이썬  1.287682


### 📋 실행 결과 데이터

| Index | 단어 | IDF 값 |
| :---: | :--- | :---: |
| **0** | **머신러닝** | 1.693147 |
| **1** | **입문** | 1.693147 |
| **2** | **데이터** | 1.287682 |
| **3** | **분석** | 1.287682 |
| **4** | **파이썬** | 1.287682 |

---

### 💡 핵심 분석 및 시사점

1. **높은 IDF (1.693...): `머신러닝`, `입문`**
   * **이유:** 전체 3개 문서 중 **단 1개 문서에만 등장**하는 희귀한 단어입니다[cite: 2].
   * **의미:** 특이성이 높으므로 문서의 개성을 나타내는 핵심 키워드로 인정받아 **더 높은 가중치(가산점)**를 부여받습니다.

2. **낮은 IDF (1.287...): `데이터`, `분석`, `파이썬`**
   * **이유:** 전체 3개 문서 중 **2개 이상의 문서에 넓게 등장**하는 흔한 단어입니다[cite: 2].
   * **의미:** 문서를 서로 구별해주는 힘이 떨어지므로 **상대적으로 낮은 가중치(감점)**를 받습니다.

3. **핵심 요약**
   * 수치 자체를 외울 필요는 없으며, **"전체 문서 집합에서 흔한 단어일수록 IDF가 낮아지고, 드문 단어일수록 IDF가 높아진다"**는 상대적 차이를 이해하는 것이 핵심입니다.

## 실습 26. 실제 도서 제목을 TF-IDF로 변환하기

In [35]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

# 실습 26. 실제 도서 제목을 TF-IDF로 변환하기

# 1. TfidfVectorizer 객체 생성 및 학습/변환
tfidf_vectorizer = TfidfVectorizer()
X_tfidf = tfidf_vectorizer.fit_transform(titles)
tfidf_terms = tfidf_vectorizer.get_feature_names_out()

# 2. 행렬 크기 및 구성 요소 출력
print("TF-IDF 행렬 크기:", X_tfidf.shape)
print("문서 수:", X_tfidf.shape[0])
print("단어 수:", X_tfidf.shape[1])

# 3. CountVectorizer와 직접 비교 (동일 데이터 기준)
print("\n--- Count vs TF-IDF 크기 비교 ---")
print("Count shape:", X_count.shape)
print("TF-IDF shape:", X_tfidf.shape)

TF-IDF 행렬 크기: (199, 536)
문서 수: 199
단어 수: 536

--- Count vs TF-IDF 크기 비교 ---
Count shape: (199, 536)
TF-IDF shape: (199, 536)


### 💻 실행 결과

```text
TF-IDF 행렬 크기: (199, 536)
문서 수: 199
단어 수: 536

--- Count vs TF-IDF 크기 비교 ---
Count shape: (199, 536)
TF-IDF shape: (199, 536)

## 실습 27. 첫 번째 도서의 TF-IDF 주요 단어 확인하기

In [36]:
import pandas as pd

# 실습 27. 첫 번째 도서의 TF-IDF 주요 단어 확인하기

# 1. 첫 번째 도서 제목 확인
print("도서 제목:", titles.iloc[0])

# 2. 첫 번째 행(희소 행렬)의 TF-IDF 값 추출
first_tfidf_row = X_tfidf.getrow(0)

# 3. 단어와 TF-IDF 값을 데이터프레임으로 생성
first_tfidf_df = pd.DataFrame({
    "단어": tfidf_terms[first_tfidf_row.indices],
    "TF-IDF": first_tfidf_row.data,
})

# 4. TF-IDF 값 기준 내림차순 정렬
first_tfidf_df = first_tfidf_df.sort_values(
    "TF-IDF",
    ascending=False,
).reset_index(drop=True)

first_tfidf_df

도서 제목: 소년이 온다


,단어,TF-IDF
0,소년이,0.707107
1,온다,0.707107


### 💻 실행 결과

* **도서 제목:** 소년이 온다[cite: 4]

| Index | 단어 | TF-IDF 값 |
| :---: | :--- | :---: |
| **0** | **소년이** | 0.707107 |
| **1** | **온다** | 0.707107 |

---

### 💡 결과 분석 및 체크리스트 답변

1. **원래 제목에 있는 단어만 나오는가?**
   * 네, 실제 제목인 **"소년이 온다"**에 등장한 단어만 추출되어 나타납니다[cite: 4].

2. **단어들의 TF-IDF 값이 동일한 이유 (`0.707107`)**
   * 두 단어(`소년이`, `온다`) 모두 해당 제목 내에서 각각 1회씩 등장하였습니다[cite: 4].
   * 전체 199개 도서 제목 집합 내에서도 두 단어의 문서 빈도(DF) 및 희소성(IDF)이 비슷하여 동일하게 높은 가중치가 부여되었고, **L2 정규화**가 적용되어 각각 $\frac{1}{\sqrt{2}} \approx 0.707107$의 값을 가집니다[cite: 3, 4].

3. **핵심 시사점**
   * 다른 흔한 단어(예: '위한', '에디션' 등)가 섞여 있지 않은 독자적인 제목 키워드이므로 두 단어 모두 문서를 대표하는 중요 피처로 동등하게 평가되었습니다[cite: 4].

## 실습 28. 여러 도서의 주요 TF-IDF 단어 확인 함수 만들기

In [37]:
import pandas as pd

# 실습 28. 여러 도서의 주요 TF-IDF 단어 확인 함수 만들기


def show_top_tfidf_terms(doc_index, top_n=5):
  # 지정한 인덱스의 TF-IDF 행 가져오기
  row = X_tfidf.getrow(doc_index)

  # 단어와 가중치로 DataFrame 생성
  result = pd.DataFrame({
      "단어": tfidf_terms[row.indices],
      "TF-IDF": row.data,
  })

  # TF-IDF 기준 상위 top_n개 정렬
  result = result.sort_values(
      "TF-IDF",
      ascending=False,
  ).head(top_n)

  print("도서 제목:", titles.iloc[doc_index])
  return result.reset_index(drop=True)


# --- 함수 실행 테스트 ---

# 1. 0번째 도서 확인
print("=== 0번째 도서 ===")
display(show_top_tfidf_terms(0, top_n=5))

# 2. 10번째 도서 확인
print("\n=== 10번째 도서 ===")
display(show_top_tfidf_terms(10, top_n=5))

# 3. 100번째 도서 확인
print("\n=== 100번째 도서 ===")
display(show_top_tfidf_terms(100, top_n=5))

=== 0번째 도서 ===
도서 제목: 소년이 온다


,단어,TF-IDF
0,소년이,0.707107
1,온다,0.707107



=== 10번째 도서 ===
도서 제목: 작별하지 않는다


,단어,TF-IDF
0,작별하지,0.752078
1,않는다,0.659074



=== 100번째 도서 ===
도서 제목: 말투만 바꿨을 뿐인데


,단어,TF-IDF
0,말투만,0.57735
1,바꿨을,0.57735
2,뿐인데,0.57735


### 💻 실행 결과 데이터

#### 1. 0번째 도서
* **도서 제목:** 소년이 온다[cite: 5]

| Index | 단어 | TF-IDF 값 |
| :---: | :--- | :---: |
| **0** | **소년이** | 0.707107 |
| **1** | **온다** | 0.707107 |

---

#### 2. 10번째 도서
* **도서 제목:** 작별하지 않는다[cite: 5]

| Index | 단어 | TF-IDF 값 |
| :---: | :--- | :---: |
| **0** | **작별하지** | 0.752078 |
| **1** | **않는다** | 0.659074 |

---

#### 3. 100번째 도서
* **도서 제목:** 말투만 바꿨을 뿐인데[cite: 5]

| Index | 단어 | TF-IDF 값 |
| :---: | :--- | :---: |
| **0** | **말투만** | 0.57735 |
| **1** | **바꿨을** | 0.57735 |
| **2** | **뿐인데** | 0.57735 |

---

### 💡 핵심 분석 및 시사점

1. **단어 수에 따른 가중치 크기 (L2 정규화 영향)**
   * **단어 2개 문장 (`소년이 온다`):** L2 정규화 공식($\sqrt{x_1^2 + x_2^2} = 1$)에 따라 각 단어가 기본적으로 $\frac{1}{\sqrt{2}} \approx 0.707$ 내외의 가중치를 갖습니다[cite: 5].
   * **단어 3개 문장 (`말투만 바꿨을 뿐인데`):** 포함된 단어가 3개로 늘어나면서 각각 $\frac{1}{\sqrt{3}} \approx 0.577$ 수준으로 가중치가 나누어집니다[cite: 5].

2. **단어 간 가중치 차이 (`작별하지 않는다`)**
   * `'작별하지'`(0.752)가 `'않는다'`(0.659)보다 높은 값을 가집니다[cite: 5].
   * 이는 전체 199개 도서 제목 집합 내에서 **`작별하지`라는 단어가 `않는다`라는 단어보다 상대적으로 덜 등장(희소성이 높음)**하여 더 높은 IDF 가중치를 받았기 때문입니다[cite: 5]

## 실습 29. 전체 데이터에서 평균 TF-IDF가 높은 단어 확인하기

In [38]:
import numpy as np
import pandas as pd

# 실습 29. 전체 데이터에서 평균 TF-IDF가 높은 단어 확인하기

# 1. 각 단어(열)별 TF-IDF 평균값 계산
mean_tfidf = np.asarray(X_tfidf.mean(axis=0)).ravel()

# 2. 단어와 평균 TF-IDF 값을 데이터프레임으로 생성
tfidf_summary = pd.DataFrame({
    "단어": tfidf_terms,
    "평균_TFIDF": mean_tfidf,
})

# 3. 평균 TF-IDF 기준 내림차순(높은 순) 정렬
tfidf_summary = tfidf_summary.sort_values(
    "평균_TFIDF",
    ascending=False,
).reset_index(drop=True)

# 4. 상위 30개 단어 출력
tfidf_summary.head(30)

,단어,평균_TFIDF
0,에디션,0.019330
1,리커버,0.013861
2,기념,0.012711
3,위한,0.012087
4,해커스,0.012072
5,토익,0.010950
6,스페셜,0.009901
7,나는,0.009786
8,내가,0.009454
9,대하여,0.008822


### 💡 핵심 분석 및 시사점

1. **상위 주요 키워드 특성:**
   * `'에디션'`, `'리커버'`, `'기념'`, `'스페셜'` 등 **특별판/한정판 관련 도서 수식어**가 전체 데이터셋에서 상당히 높은 평균 TF-IDF 가중치를 차지합니다[cite: 6].
   * `'해커스'`, `'토익'` 등 특정 분야(수험서/자격증)에 집중 등장하면서 개성이 뚜렷한 키워드가 뒤를 잇고 있습니다[cite: 6].

2. **수치 해석 시 유의점:**
   * **단순 평균의 특징:** 전체 199개 문서 중 대부분의 문서에서 등장하지 않은 단어는 값이 `0`이므로, 평균 수치 자체는 `0.01` 내외로 작아 보입니다[cite: 6].
   * **원인 해석 금지:** `'에디션'`의 평균 TF-IDF가 가장 높다고 해서 이 단어가 베스트셀러가 된 **직접적인 원인**이라고 해석해서는 안 됩니다. 현재 수집된 전체 제목 데이터 안에서 상대적으로 수치적 영향력이 높게 측정된 특징량(Feature)일 뿐입니다[cite: 6].

## 실습 30. TF-IDF 상위 단어 저장하기

In [39]:
import pandas as pd

# 실습 30. TF-IDF 상위 단어 저장하기

# 1. 평균 TF-IDF 상위 30개 단어 추출
tfidf_top30 = tfidf_summary.head(30)

# 2. CSV 파일로 저장 (한글 깨짐 방지를 위해 utf-8-sig 사용)
tfidf_top30.to_csv(
    "chapter03_tfidf_top_terms.csv",
    index=False,
    encoding="utf-8-sig",
)

# 3. 저장된 CSV 파일을 다시 불러와 확인
df_tfidf_check = pd.read_csv(
    "chapter03_tfidf_top_terms.csv",
    encoding="utf-8-sig",
)

# 4. 상위 5개 행 출력
df_tfidf_check.head()

,단어,평균_TFIDF
0,에디션,0.019330
1,리커버,0.013861
2,기념,0.012711
3,위한,0.012087
4,해커스,0.012072


### 💡 핵심 요약 및 결과 분석

1. **정상 저장 및 복원 확인:**
   * `chapter03_tfidf_top_terms.csv` 파일로 저장했던 상위 단어 데이터가 인코딩 오류(한글 깨짐) 없이 정상적으로 불러와짐을 확인했습니다[cite: 7].
   * `index=False` 옵션을 지정했기 때문에 불필요한 인덱스 열 생성 없이 `단어`와 `평균_TFIDF` 2개 컬럼만 정상적으로 로드되었습니다[cite: 7].

2. **상위 5개 키워드 특성:**
   * 도서 제목 데이터셋 전체에서 **'에디션'**, **'리커버'**, **'기념'** 등 한정판/특별판 관련 어휘와 **'위한'**, **'해커스'** 같은 주요 목적/출판사 어휘가 상위 가중치를 형성하고 있습니다[cite: 7].

## 31. Count 상위 단어와 TF-IDF 상위 단어 비교하기

In [40]:
import pandas as pd

# 실습 31. Count 상위 단어와 TF-IDF 상위 단어 비교하기

# 1. 두 결과(Count 상위 30, TF-IDF 상위 30)를 하나의 DataFrame으로 결합
comparison = pd.DataFrame({
    "Count_상위단어": count_top30["단어"].reset_index(drop=True),
    "Count_전체등장횟수": count_top30["전체등장횟수"].reset_index(drop=True),
    "TFIDF_상위단어": tfidf_top30["단어"].reset_index(drop=True),
    "평균_TFIDF": tfidf_top30["평균_TFIDF"].reset_index(drop=True),
})

# 2. 상위 20개 비교 출력
comparison.head(20)

,Count_상위단어,Count_전체등장횟수,TFIDF_상위단어,평균_TFIDF
0,에디션,12,에디션,0.019330
1,리커버,8,리커버,0.013861
2,기념,8,기념,0.012711
3,위한,7,위한,0.012087
4,해커스,7,해커스,0.012072
5,토익,6,토익,0.010950
6,나는,5,스페셜,0.009901
7,내가,5,나는,0.009786
8,스페셜,5,내가,0.009454
9,싶은,4,대하여,0.008822


### 💡 핵심 비교 및 해석 정리

1. **상위 순위 변동 단어 (예: `스페셜`)**
   * **Count 기준:** 8위 (5회 등장)[cite: 8]
   * **TF-IDF 기준:** 6위로 상승 (0.009901)[cite: 8]
   * **원인:** `'스페셜'`이라는 단어가 등장하는 제목들 안에서 문서 식별력이 높아 TF-IDF 가중치가 상대적으로 높게 평가되었습니다[cite: 8].

2. **순위 하락 단어 (예: `나는`, `내가`)**
   * **Count 기준:** 각각 6위, 7위 (5회 등장)[cite: 8]
   * **TF-IDF 기준:** 각각 7위, 8위로 순위 하락[cite: 8]
   * **원인:** 여러 도서 제목에 범용적으로 등장하여 문서 빈도(DF)가 상대적으로 높고, 이에 따라 IDF 감점을 받았기 때문입니다[cite: 8].

3. **두 방식의 핵심 차이점**
   * **CountVectorizer:** 전체 문서에서 '몇 번 나왔는가(등장 횟수)'만을 단순 측정합니다[cite: 8].
   * **TfidfVectorizer:** 해당 문서에서의 등장 빈도(TF)와 함께 **전체 문서 집합에서의 희소성(IDF)**을 반영하여, 흔한 단어는 감점하고 특정 문서를 대표하는 독자적인 단어에 더 높은 가중치를 부여합니다[cite: 8].

## 실습 47. 일부 도서 결과를 직접 검증하기

In [41]:
# 실습 47. 일부 도서 결과를 직접 검증하기

# 1. 검증할 도서의 인덱스 번호 지정
sample_indices = [0, 10, 20]

# 2. 지정한 인덱스의 도서 제목과 TF-IDF 상위 단어 확인
for index in sample_indices:
  if index < len(titles):
    print("=" * 60)
    print("도서 제목:", titles.iloc[index])
    display(show_top_tfidf_terms(index, top_n=5))

도서 제목: 소년이 온다
도서 제목: 소년이 온다


,단어,TF-IDF
0,소년이,0.707107
1,온다,0.707107


도서 제목: 작별하지 않는다
도서 제목: 작별하지 않는다


,단어,TF-IDF
0,작별하지,0.752078
1,않는다,0.659074


도서 제목: 쇼펜하우어 인생수업(30만 부 기념 개정증보판)
도서 제목: 쇼펜하우어 인생수업(30만 부 기념 개정증보판)


,단어,TF-IDF
0,개정증보판,0.498481
1,쇼펜하우어,0.462422
2,인생수업,0.462422
3,30만,0.436838
4,기념,0.364720


### 💡 주요 검증 질문별 시사점 및 분석

1. **상위 단어가 실제 제목에 존재하는가?**
   * 네, 추출된 모든 상위 피처는 각 도서 제목에 포함된 어휘들로 정확히 일치합니다[cite: 9].

2. **숫자나 기호의 처리 상태 (`30만 부 기념`)**
   * 기본 Vectorizer 토큰화 설정으로 인해 **`30만`** 같은 숫합형 단어가 그대로 피처로 추출되었습니다[cite: 9]. 
   * 괄호 등의 특수문자는 토큰화 과정에서 자동 분리 및 제거되었습니다[cite: 9].

3. **흔한 단어의 가중치 하락 (`기념`)**
   * `'기념'`(0.364720)은 전체 베스트셀러 제목들에서 상대적으로 자주 등장하는 단어이므로, 동일한 문장 안의 희귀 어휘인 `'개정증보판'`(0.498481)이나 `'쇼펜하우어'`(0.462422)보다 낮은 TF-IDF 수치를 가집니다[cite: 9].

4. **형태소 분석기 도입의 필요성**
   * 현재는 기본 공백/기호 분할 기준이므로 `'소년이'`, `'작별하지'`처럼 조사 및 어미가 결합된 형태 그대로 피처가 구성됩니다[cite: 9].
   * Kiwi나 Okt 같은 한국어 형태소 분석기를 도입하여 명사 단위(`소년`, `작별`, `쇼펜하우어`, `인생`, `수업` 등)로 정제하여 입력을 전달하면 단어의 의미적 대표성이 훨씬 더 향상될 것입니다[cite: 9].

## 실습 50. 왜 Chapter 04에서는 전체 데이터에 먼저 fit하면 안 될까?

### 📌 데이터 누수(Data Leakage) 발생 메커니즘

| 구분 | 올바른 접근법 (Train/Test 분리 후 fit) | 잘못된 접근법 (전체 데이터에 fit) |
| :---: | :--- | :--- |
| **단어 사전 구축** | Train 데이터에 포함된 단어만 수집하여 사전을 형성함 | Test 데이터에만 등장하는 단어 정보까지 사전에 미리 포함됨 |
| **IDF 값 계산** | Train 데이터 내의 문서 빈도(DF)만을 기준으로 가중치 산출 | Test 데이터의 문서 빈도까지 포함되어 IDF 수치가 왜곡됨 |
| **모델 평가** | 미지의 실제 데이터에 대한 일반화 성능을 정확히 평가 가능 | 모델 성능이 실제보다 과도하게 좋게 나오는 **낙관적 편향** 발생 |

---

### 🛠️ 올바른 파이프라인 흐름 (Chapter 04 적용 방식)

```python
# 1. Train / Test 데이터셋 먼저 분리
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# 2. Train 데이터로만 단어 사전 및 IDF 학습 (fit)
vectorizer = TfidfVectorizer()
X_train_tfidf = vectorizer.fit_transform(X_train)  # fit + transform

# 3. Test 데이터는 학습된 사전과 IDF를 바탕으로 변환만 수행 (transform)
X_test_tfidf = vectorizer.transform(X_test)  # fit 금지! transform만 실행

In [43]:
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split

# 1. 데이터 불러오기
df_books = pd.read_csv("book_bestseller_clean.csv", encoding="utf-8-sig")

# 2. X(상품명) 데이터 전처리 및 추출
X = df_books["상품명"].fillna("").astype(str).str.strip()

# 3. y 없이 X만 Train/Test 분리 (8:2 비율)
X_train, X_test = train_test_split(X, test_size=0.2, random_state=42)

print(f"Train 개수: {len(X_train)}")
print(f"Test 개수: {len(X_test)}")

# 4. TfidfVectorizer 정의 및 Train fit/transform
tfidf_vectorizer = TfidfVectorizer()
X_train_tfidf = tfidf_vectorizer.fit_transform(X_train)

# 5. Test transform 수행 (fit 금지)
X_test_tfidf = tfidf_vectorizer.transform(X_test)

print("\n--- TF-IDF 행렬 크기 ---")
print("X_train_tfidf shape:", X_train_tfidf.shape)
print("X_test_tfidf shape:", X_test_tfidf.shape)

Train 개수: 159
Test 개수: 40

--- TF-IDF 행렬 크기 ---
X_train_tfidf shape: (159, 435)
X_test_tfidf shape: (40, 435)


## 실습 51. Chapter 04에서 사용할 올바른 순서 미리 보기

---

### 🔄 전체 데이터 처리 흐름 (Pipeline)

```text
1. 원본 데이터 준비 (CSV 파일 로드)
       ↓
2. 피처(X, 도서 제목)와 레이블(y, 카테고리/분류) 설정
       ↓
3. Train 데이터셋과 Test 데이터셋으로 분리 (train_test_split)
       ↓
4. TfidfVectorizer를 Train 데이터에만 fit() 실행 (단어 사전 및 IDF 학습)
       ↓
5. Train 데이터를 수치 벡터로 변환 (fit_transform)
       ↓
6. Test 데이터는 학습된 사전 기반으로 변환만 수행 (transform 만 실행!)
       ↓
7. 나이브 베이즈(Naive Bayes) 등 분류 모델 학습 (fit)
       ↓
8. Test 데이터 예측 및 성능 평가 (predict & accuracy_score)